### PR TITLE
Warn the user if they've activated the wrong major-mode

### PR DESCRIPTION
**Pull request by Malabarba**
https://github.com/clojure-emacs/cider/issues/1611

### CODE REVIEW

File: clojure-mode.el [hunk 1]
```diff
@@ -329,6 +329,33 @@ instead of to `clojure-mode-map'."
   (clojure-font-lock-setup)
   (add-hook 'paredit-mode-hook #'clojure-paredit-setup))
 
+(defcustom clojure-do-warnings t
```
Comment (bbatsov) [inline clojure-mode.el]: I'd name this differently - e.g. `clojure-verify-major-mode`. How do you even do warnings? :-)

Comment (Malabarba) [inline clojure-mode.el]: Fair enough. 😝

File: clojure-mode.el [hunk 2]
```diff
@@ -329,6 +329,36 @@ instead of to `clojure-mode-map'."
...
+    (let ((problem
+           (cond ((and (string-match "\\.cljs\\'" (buffer-file-name))
+                       (not (eq major-mode 'clojurescript-mode)))
+                  'clojurescript-mode)
+                 ((and (string-match "\\.clj[cx]\\'" (buffer-file-name))
+                       (not (eq major-mode 'clojurec-mode)))
+                  'clojurec-mode))))
+      (when problem
+        (message "[WARNING] %s activated `%s' instead of `%s' in this buffer.
+This is could cause problems.
```
Comment (bbatsov) [inline clojure-mode.el]: This could cause problems.

File: clojure-mode.el [hunk 3]
```diff
@@ -329,6 +329,36 @@ instead of to `clojure-mode-map'."
...
+  "Check if the current major-mode matches the file extension.
+If it doesn't, issue a warning if `clojure-verify-major-mode' is
+non-nil."
+  (when (and clojure-verify-major-mode
+             (stringp (buffer-file-name)))
+    (let ((problem
+           (cond ((and (string-match "\\.cljs\\'" (buffer-file-name))
+                       (not (eq major-mode 'clojurescript-mode)))
+                  'clojurescript-mode)
+                 ((and (string-match "\\.clj[cx]\\'" (buffer-file-name))
```
Comment (bbatsov) [inline clojure-mode.el]: The mode for cljx is different.

File: clojure-mode.el [hunk 4]
```diff
@@ -329,6 +329,36 @@ instead of to `clojure-mode-map'."
   (clojure-font-lock-setup)
   (add-hook 'paredit-mode-hook #'clojure-paredit-setup))
 
+(defcustom clojure-verify-major-mode t
+  "If non-nil, `clojure-mode' will warn of possible user errors."
```
Comment (bbatsov) [inline clojure-mode.el]: Maybe you should update this docstring now?

**Comment by bbatsov** [comment]
Apart from my small remarks - this needs a changelog entry. Otherwise - good idea!

**Pull request merged**